crates/metrics/src/typed_prometheus.rs

```diff
@@ -71,17 +78,34 @@ pub use make_collector;
 macro_rules! metrics_histogram_vec {
     ($vis:vis $name:ident: $vecty:ident($($labels:ident: $labelty:ty),+ $(,)?) ($($bucket:literal)*)) => {
         #[derive(Clone)]
-        $vis struct $name($vecty);
+        $vis struct $name($vecty, tokio::sync::mpsc::UnboundedSender<Box<dyn FnOnce() + Send>>);
         impl $name {
-            pub fn with_opts(opts: prometheus::Opts) -> prometheus::Result<Self> {
+            pub fn with_opts(opts: prometheus::Opts, tx: tokio::sync::mpsc::UnboundedSender<Box<dyn FnOnce() + Send>>) -> prometheus::Result<Self> {
                 let opts = prometheus::HistogramOpts::from(opts).buckets(vec![$(f64::from($bucket)),*]);
-                $vecty::new(opts.into(), &[$(stringify!($labels)),+]).map(Self)
+                Ok(Self($vecty::new(opts.into(), &[$(stringify!($labels)),+] ).unwrap(), tx))
             }
 
             pub fn with_label_values(&self, $($labels: &$labelty),+) -> <$vecty as $crate::typed_prometheus::ExtractMetricVecT>::M {
                 use $crate::typed_prometheus::AsPrometheusLabel as _;
                 self.0.with_label_values(&[ $($labels.as_prometheus_str().as_ref()),+ ])
             }
+
+            pub fn with_label_values_async<F>(&self, $($labels: &$labelty),+, op: F)
+            where
+                F: FnOnce(
+                        <$vecty as $crate::typed_prometheus::ExtractMetricVecT>::M,
+                    ) + Send + 'static,
+            {
+                use $crate::typed_prometheus::AsPrometheusLabel as _;
+                let this = self.clone();
+                $(
+                    let $labels = $labels.to_owned();
+                )+
+                let _ = self.1.send(Box::new(move || {
```

Comment:
I'm not sure about this allocation, I guess let's see what the benchmarks tell us.